src/Symfony/Component/DependencyInjection/Exception/ServiceNotFoundException.php

```diff
@@ -12,45 +12,16 @@
 namespace Symfony\Component\DependencyInjection\Exception;
 
 /**
- * This exception is thrown when a non-existent service is requested.
+ * Abstract exception extended by both RequestedServiceNotFoundException and MissingDependencyException.
  *
  * @author Johannes M. Schmitt <schmittjoh@gmail.com>
  */
-class ServiceNotFoundException extends InvalidArgumentException
+abstract class ServiceNotFoundException extends InvalidArgumentException
 {
-    private $id;
-    private $sourceId;
-
-    public function __construct($id, $sourceId = null, \Exception $previous = null, array $alternatives = array())
-    {
-        if (null === $sourceId) {
-            $msg = sprintf('You have requested a non-existent service "%s".', $id);
-        } else {
-            $msg = sprintf('The service "%s" has a dependency on a non-existent service "%s".', $sourceId, $id);
-        }
-
-        if ($alternatives) {
-            if (1 == count($alternatives)) {
-                $msg .= ' Did you mean this: "';
-            } else {
-                $msg .= ' Did you mean one of these: "';
-            }
-            $msg .= implode('", "', $alternatives).'"?';
-        }
-
-        parent::__construct($msg, 0, $previous);
-
-        $this->id = $id;
-        $this->sourceId = $sourceId;
-    }
+    protected $id;
 
     public function getId()
     {
         return $this->id;
     }
-
-    public function getSourceId()
```

Comment:
@moufmouf yes to both questions At least that's the default answer, and if one thinks otherwise, one must demonstrates why at the technical level. Here, I doubt this is possible. So: yes and yes.